### PR TITLE
Select CompileParams for TMA operations automatically at compile-time

### PR DESCRIPTION
**Pull request by rdspring1**
When using TMA operations, we must use 32-bit indexing and disable magic zero support. Currently, we manually pass the correct `CompileParams` to `FusionExecutor::compileFusion`. This PR automatically updates the `CompileParams` if TMA operations are detected during compilation. It also includes a basic normalization kernel using TMA operations in the python frontend.

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

File: csrc/python_frontend/fusion_definition.cpp [hunk 1]
```diff
@@ -145,11 +146,16 @@ void FusionDefinition::finalizeSchedule(
...
   if (multidevice_executor_ == nullptr) {
+    // NOTE: TMA operations require 32-bit indexing and magic zero support to be
+    // disabled.
     user_sched_->executor->compileFusion(
         user_sched_->schedule.get(),
         inputs,
         user_sched_->fusion_id_,
-        user_sched_->device_id_);
+        user_sched_->device_id_,
+        (uses_tma_ops)
```
Comment (jjsjann123) [inline csrc/python_frontend/fusion_definition.cpp:156]: this feels a bit hacky here.

IIUC, we should be able to traverse the fusion IR `user_sched_->schedule` looking for `LoadStoreOpType::CpAsyncBulkTensorTile` to figure out that we should have use the modified compile param?

Comment (jjsjann123) [reply csrc/python_frontend/fusion_definition.cpp:156]: Maybe it's better to do that inside compileFusion instead?

**Review by jjsjann123 — COMMENTED** [review]
> When using TMA operations, we need to use 32-bit indexing and to disable magic zero support in the CompileParams

I have a question regarding this statement above. If this is a general thing, we probably want to have it done automatically in `compileFusion`, instead of require this as a user flag to enable it.

If it's not a general thing, plumbing it through the API seems to be a bad idea to begin with...

File: csrc/python_frontend/fusion_definition.cpp (hunk 1 above)
Comment (rdspring1) [reply csrc/python_frontend/fusion_definition.cpp:156]: We can update `CompileParam` in `compileFusion` in a separate PR and leave this one open as an example for now.

Comment (jjsjann123) [reply csrc/python_frontend/fusion_definition.cpp:156]: Sounds good. I don't have any more concerns regarding this PR. But let's chat offline and discuss how we would want to merge this stack.

**Comment by jjsjann123** [comment]
I think this is the last PR in the stack, let me know when it's ready to be reviewed.

**Comment by rdspring1** [comment]
!build

**Comment by rdspring1** [comment]
!build

File: csrc/executor.cpp [hunk 2]
```diff
@@ -343,13 +352,21 @@ void FusionExecutor::compileFusion(
...
+            (compile_params.index_type.value() == PrimDataType::Int32),
+        "Compilation with int64 is requested but int32 is required because of TMA operations");
+
   } else if (arg_index_type == PrimDataType::Int) {
     // If the given compile option doesn't specify the index type, and
     // the arguments require 64-bit indexing, we need to use 64-bit
     // indexing. Note that if the arg type is 32-bit, it doesn't mean
     // it's safe to use 32-bit for the whole kernel, so unless it's
     // specified through CompileParams, we do not use 32-bit indexing.
     compile_params.index_type = arg_index_type;
```
Comment (zasdfgbnm) [inline csrc/executor.cpp:366]: Should we add a check `NVF_ERROR(! has_cp_async_bulk_` here?

**Comment by rdspring1** [comment]
!build

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**